src/foam/dao/SyncDAO.js

```diff
@@ -112,6 +118,18 @@ foam.CLASS({
       factory: function() { return Promise.resolve(); }
     },
     {
+      name: 'versionedOf_',
+      expression: function(of) {
+        return this.versionedClassFactory_.get(of);
+      }
```

Comment:
I think what's making this seem complicated is the inclusion of the versioning adapter here. If a DAO decorator was adapting the instances on the client end of SyncDAO, then everything in here could be implemented as before, and a facade or factory class could configure the parts of the sync system to use the versioned class.